### PR TITLE
SILGen: Account for abstraction differences when assigning in key path setters

### LINKED ISSUE #63682: Runtime crash when reassigning generic closure via KeyPath
**Description**

Running the following code causes the program to crash. It seems like it has something to do with the combination of closures and associatedtypes, since accessing `value` after changing it via KeyPaths do just work fine.
```swift
protocol Abstract {
  associatedtype T
  var value: T { get }
}

struct Concrete: Abstract {
  var value: String = ""
}

struct Container<A: Abstract> {
  var closure: () -> A.T?
  var value: A.T?
}

func crash() {
  var container = Container<Concrete>(closure: { "" })
  container[keyPath: \.closure] = { "" }
  container[keyPath: \.value] = ""
  container.value
  container.closure()
}

crash()
```
```
Stack dump:
0.	Program arguments: /Applications/Xcode-14.2.0.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swift-frontend -frontend -interpret Crash.swift -enable-objc-interop -stack-check -sdk /Applications/Xcode-14.2.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk -color-diagnostics -new-driver-path /Applications/Xcode-14.2.0.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swift-driver -empty-abi-descriptor -resource-dir /Applications/Xcode-14.2.0.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift -module-name Crash -disable-clang-spi -target-sdk-version 13.1
1.	Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51)
2.	Compiling with the current language version
3.	While running user code "Crash.swift"
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x000000010fbf9fe7 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 39
1  swift-frontend           0x000000010fbf9018 llvm::sys::RunSignalHandlers() + 248
2  swift-frontend           0x000000010fbfa600 SignalHandler(int) + 288
3  libsystem_platform.dylib 0x00007ff810f3ac1d _sigtramp + 29
4  libsystem_platform.dylib 0x00007ff7b58d0a90 _sigtramp + 18446744072176098960
5  libsystem_platform.dylib 0x00000001133021f8 _sigtramp + 18446603374913418744
6  libsystem_platform.dylib 0x000000011330200e _sigtramp + 18446603374913418254
7  swift-frontend           0x000000010b9319c0 llvm::orc::runAsMain(int (*)(int, char**), llvm::ArrayRef<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, llvm::Optional<llvm::StringRef>) + 1856
8  swift-frontend           0x000000010aa40f28 swift::RunImmediately(swift::CompilerInstance&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, swift::IRGenOptions const&, swift::SILOptions const&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >&&) + 14104
9  swift-frontend           0x000000010a7b0482 performCompileStepsPostSILGen(swift::CompilerInstance&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >, llvm::PointerUnion<swift::ModuleDecl*, swift::SourceFile*>, swift::PrimarySpecificPaths const&, int&, swift::FrontendObserver*) + 3138
10 swift-frontend           0x000000010a7b3750 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 11792
11 swift-frontend           0x000000010a747eaa swift::mainEntry(int, char const**) + 3082
12 dyld                     0x00007ff810bdf310 start + 2432
```

**Environment**
- Swift compiler version info swift-driver version: 1.62.15 Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51) Target: x86_64-apple-macosx13.0
- Xcode version info Xcode 14.2 Build version 14C18
- Deployment target: iOS 16.2

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
Resolves #63682

### CODE REVIEW

File: lib/SILGen/SILGenExpr.cpp [hunk 1]
```diff
@@ -3239,6 +3239,15 @@ static SILFunction *getOrCreateKeyPathSetter(SILGenModule &SGM,
...
                         /*index for diags*/ nullptr);
 
+  // If the assigned value will need to be reabstracted, add a reabstraction
+  // component.
+  const auto loweredSubstType = subSGF.getLoweredType(lv.getSubstFormalType());
+  if (lv.getTypeOfRValue() != loweredSubstType.getObjectType()) {
+    // Logical components always re-abstract back to the substituted type.
+    assert(lv.isLastComponentPhysical());
+    lv.addOrigToSubstComponent(loweredSubstType);
+  }
```
Comment (AnthonyLatsis) [inline lib/SILGen/SILGenExpr.cpp:3249]: This logic is duplicated along several non-intersecting code paths. It looks like we should be checking for this whenever we finish assembling an `LValue`, which suggests some sort of an `LValueBuilder`.

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

**Comment by AnthonyLatsis** [comment]
@slavapestov ping

I’m hoping for this to board the release if possible.

**Review by slavapestov — APPROVED** [review]

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
Much obliged!

**Pull request merged**